programs/system/src/lib.rs

```diff
@@ -30,66 +97,161 @@ pub fn entrypoint(
-                        );
-                    } else {
-                        info!(
-                            "CreateAccount: invalid argument; account {} already in use",
-                            keyed_accounts[to].unsigned_key()
-                        );
-                    }
-                    Err(ProgramError::InvalidArgument)?;
-                }
-                if tokens > keyed_accounts[from].account.tokens {
-                    info!(
-                        "CreateAccount: insufficient tokens ({}, need {})",
-                        keyed_accounts[from].account.tokens, tokens
-                    );
-                    Err(ProgramError::ResultWithNegativeTokens)?;
-                }
-                keyed_accounts[from].account.tokens -= tokens;
-                keyed_accounts[to].account.tokens += tokens;
-                keyed_accounts[to].account.owner = program_id;
-                keyed_accounts[to].account.userdata = vec![0; space as usize];
-                keyed_accounts[to].account.executable = false;
+                create_system_account(keyed_accounts, tokens, space, program_id).map_err(
+                    |e| match e {
+                        SystemError::ResultWithNegativeTokens => {
+                            ProgramError::ResultWithNegativeTokens
+                        }
+                        _ => ProgramError::InvalidArgument,
+                    },
+                )
             }
             SystemInstruction::Assign { program_id } => {
-                if !system_program::check_id(&keyed_accounts[from].account.owner) {
-                    Err(ProgramError::AssignOfUnownedAccount)?;
-                }
-                keyed_accounts[from].account.owner = program_id;
-            }
-            SystemInstruction::Move { tokens } => {
-                let to = 1;
-
-                // bank should be verifying correctness
-                if tokens > keyed_accounts[from].account.tokens {
-                    info!(
-                        "Move: insufficient tokens ({}, need {})",
-                        keyed_accounts[from].account.tokens, tokens
-                    );
-                    Err(ProgramError::ResultWithNegativeTokens)?;
-                }
-                keyed_accounts[from].account.tokens -= tokens;
-                keyed_accounts[to].account.tokens += tokens;
+                assign_account_to_program(keyed_accounts, program_id)
             }
+            SystemInstruction::Move { tokens } => move_tokens(keyed_accounts, tokens),
         }
-        Ok(())
     } else {
         info!("Invalid transaction instruction userdata: {:?}", data);
-        Err(ProgramError::InvalidArgument)
+        Err(ProgramError::InvalidUserdata)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use solana_sdk::account::Account;
+    use solana_sdk::signature::{Keypair, KeypairUtil};
+
+    #[test]
+    fn test_create_system_account() {
```

Comment:
This looks like 4 independent tests.  Why stuff them together?

Reply:
No good reason. They are split now.